### ISSUE TITLE
Proposal: Define our goals and roadblocks to release a 3.0 beta

### ISSUE THREAD

**Issue by Robbt** [libretime/libretime#544]
Basically there has been discussion at various place about when we can transition LibreTime from alpha to beta. I think that we need to explicitly call out any show stopper bugs or missing functionality that would prevent us from transitioning to a beta vs. alpha release.

Here is the list of bugs I've compiled and originally posted as comments below.
## Blocker issues for LT beta  ##
- [x] #552 Automate silan fix during install 
- [ ] #42 Line-in recording
- [x] #317 ensure locale works - in theory fixed by python 3 support
- [x] #129 autoplaylist depends on API calls to run
- [x] #461 issue with web player widget not working in certain browsers 
- [x] #71 - either have an automatic upgrade path from airtime 2.5.1 + 2.5.2 or provide documentation for a tested approach that allows people to try the upgrade w/o data loss and report errors
- [x] #70 Watched Folders and/or CLI import
- [x] #550 Media monitor config 
- [x] #577 ensure mediafolders/watched folders configuration doesn't break LT when upgrading to LT from AT
- [x] #580 we need to have a system that works with modern distros that use PHP 7.2
- [x] #376 - address all glaring issues in the docs where we refer to broken or non-existent features and ensure all introduced features have reasonable documentation

- [ ] #624 Provide an easy installation process that is supported aside from the install script - this could be default docker containers or debian packages or both and explicitly define what distros we support out of the box and which ones are experimental and unsupported
- [x] #623 Provide a documented way of upgrading the software to the latest version
- [x] #190 Provide a well documented way to backup a LT instance 
- [x] #951 Merge python3 support

Since this is a subjective call we get to define the criteria. Up until now releases have been done at the discretion of @hairmare and they've been well documented and put together. 

When consulting the above list also look at the related [beta 3 blocker project](https://github.com/LibreTime/libretime/projects/2) (none are authoritative atm).

**Comment by hairmare** [comment]
I see the known issues section of the release notes as containing the main blockers for a beta. IMO it contains the issues needing a fix so all existing legacy upstream users have a way to transition to LibreTime without missing out on any features.

**Comment by hairmare** [comment]
Adding features before we reach beta is fine with me, waiting for 3.1.0 would be preferable but getting people to hold off implementing features they need doesn't seem pragmatic.

IMO the most important issues will be things like watched folders and line-in recording should have the highest priority since they will block folks updating from legacy upstream.

I think milestones are preferable to labels for tagging any 3.0.0 issues, an additional blocker label might make sense so we can use `is:issue milstone:3.0.0 label:release-blocker` to filter the issues.

**Comment by Robbt** [comment]
Ok, here is a list issues mentioned in the release notes. Feel free to add other bugs in comments that you consider show stoppers before we do a beta.

- [ #70 Watched Folders ]
- [ #550 Media monitor config ]
- [ #552 Automate silan fix during install ]
- [ #42 Line-in recording]
- [ #553 ensure local works ]
- [#301 i8 support ]

**Comment by Robbt** [comment]
We should also tag those as @hairmare recommended. I think the Silan fix would be great to automate. I'm not sure that we want to support watched folders but do think some kind of file system import is necessary. Also not sure if we should consider line-in recording an essential task since nobody appears to be working on it. The Silan and media monitor can probably be fixed by the install script.

**Comment by frecuencialibre** [comment]
For us, the line-in recording is in the nice-to-have category. automatic playlists that "just work" as one would expect from the UI (ie. without the confusing need to have set them up at least an hour beforehand and have had active track changes during that time) are top priority. I know I discouraged feature requests, but since this one is already in the UI, and also frequently requested, perhaps we can include it. @Robbt sorry for the possibly obvious question, but would your [proposed solution](https://github.com/LibreTime/libretime/issues/129#issuecomment-294267272) cause automatic playlists to work immediately? If so, I'd humbly request that it be prioritized for beta.

Also, I'll also add one more issue to our list: documentation: https://github.com/LibreTime/libretime/issues/376 I plan on continuing to chip away here, and see someone else also volunteered but perhaps just needs some company.

**Comment by ned-kelly** [comment]
> We should also tag those as @hairmare recommended. I think the Silan fix would be great to automate. I'm not sure that we want to support watched folders but do think some kind of file system import is necessary. Also not sure if we should consider line-in recording an essential task since nobody appears to be working on it. The Silan and media monitor can probably be fixed by the install script.

Filesystem import is critical for a beta release IMHO - Any station looking to adopt Libretime from another system is going to want a quick and easy to just copy files in without having to worry about uploading through the UI etc..

**Comment by JohnnyC1951** [comment]
Add some more common dependencies (show stoppers) missing from some VPS distros, especially OVZ ones. I sent @Robbt a short & simple list by email that recently choked an install on an OVZ server along with a reinstall 'gotcha' fix that really needs adding to the install script. If the script fails, it wont reinstall properly as airtime.conf is already airtime.tmp, so I suggest check for either existing.

**Comment by Robbt** [comment]
Hey @JohnnyC1951 I sent you an e-mail back, I was unable to find the packages you suggested were needed under trusty and they are already included under xenial (except for the software-properties-common) and thought I found an issue with the installer not enabling rewrite but realized that is not the case. Software-properties-common is used for adding new repositories and so installing this will be included if we start installing non-distro packages by default and doesn't need to be in our current install until that happens.

I can open an issue for the airtime.tmp issue as this does seem like a bug in terms of a failure to recover from a broken state with the installer script.

**Comment by Robbt** [comment]
With regards to the auto playlist I added the issue where it depends upon track or API #129  calls to run to the blocking bugs. 
With regards to whether an auto playlist should schedule whenever a new show is added that contains the playlist this could probably be done somewhat easily but this will only typically occur when someone is trying to figure out how automatic playlists work and exploring the software as people aren't usually creating shows that air right now (except when they are just trying to see how libretime works) and instead are creating shows for the future. But you aren't the first person to think "huh autoplaylists doesn't work" because it doesn't do anything immediately. So there is a need to make the model more understandable, explain it better and maybe modify the way it works for special cases like trying a autoplaylist check when a show is created. But we should create a separate issue for this.

**Comment by JohnnyC1951** [comment]
@Robbt Epistle sent :\

I DO NOT WANT Auto-Playlists populating at add-to-show time. It screws up my 'just in time' loading of the very latest content. Just my 2c.
Just add a paragraph in the manual. ;)

**Comment by Robbt** [comment]
Turns out #129 isn't a problem, I'm not sure what was causing the issue in the past but recent testing confirms that indeed LibreTime will hit the API every 8 minutes and trigger the various tasks even when nothing is scheduled or longer shows are being played. So I knocked that one off the todo list.

I'd say we should make sure that we solve #71 and have a clear upgrade path from 2.5.x and also #376 so that we don't have any glaringly incorrect documentation. If we can do both of these and perhaps even get some debian packages together then I'd say we could make the transition to calling LibreTime a beta.

**Comment by Robbt** [comment]
So I updated the checklist above to consolidate our work on the beta and come up with an idealistic list of what I think would be important for us to call ourselves a beta. I think that more and more people are going to be coming over from AT 2.5.1 as they realize it is essentially EOL and can't run on a modern distro.

**Comment by gusaus** [comment]
> I think milestones are preferable to labels for tagging any 3.0.0 issues, an additional blocker label might make sense so we can use `is:issue milstone:3.0.0 label:release-blocker` to filter the issues.

Incorportating this suggestion from @hairmare https://github.com/LibreTime/libretime/issues/544#issuecomment-429404462 would also make it easier for new contributors to find issues to work on. Are there other issues that should be added to the list in the issue summary?

**Comment by gusaus** [comment]
It might be out of context for this issue, but relaunching the libretime.org website by the beta would  https://github.com/LibreTime/libretime/pull/590 provide a much better showcase for users and contributors.

**Comment by Robbt** [comment]
I agree, I'll add #590 to the list. I don't think the blocker label and milestone is a bad idea and I might do it when I go through the issues again and triage them.

**Comment by gusaus** [comment]
Might also help new and/or non-developer contributors if relevent issues had labels like _help wanted_, _documentation_, _good first issue_. Many of these issues are already labeled as such.

![justjazz 2019-01-18 12-48-09](https://user-images.githubusercontent.com/113996/51403724-4fe73780-1b1f-11e9-8417-4550b0355734.png)

**Comment by Robbt** [comment]
both help wanted and documentation already have labels and some things have been labelled with them I can't think of any of the issues that would necessarily make a "good first issue" because the system is rather complex and nothing pops out as easy to solve or not requiring the ability to learn a specific programming language.

**Comment by hairmare** [comment]
I can add some "good first issue" issues if we want to. 

One of them could be getting rid of any `require_*()` or `include_*()` statments in the php parts (most of them can get replaced with composer autoloading).

We could also tag some things that revolve around working on the docs as "good first issue".

My experience with "good first issue" issues is that they need someone to take care of mentoring those who would like to contribute. With this type of issues the mentoring is mostly about using GitHub and best-practices regarding to coding. So not something I'm sure we need more of.

**Comment by gusaus** [comment]
Going on the assumption 3.0 is the next major milestone, would it make sense to set this up as a goal on the LibreTime collective? Skimming through https://github.com/LibreTime/libretime/issues/796#issuecomment-484676469, it seems like there was some agreement here. 

https://github.com/LibreTime/libretime/projects/2
https://github.com/LibreTime/libretime/milestone/9
https://docs.opencollective.com/help/collectives/goals

**Comment by gusaus** [comment]
@hairmare @Robbt So it looks like we should remove 'beta' from this issue title!?! https://github.com/LibreTime/libretime/issues/544#issuecomment-511540941

**Comment by gusaus** [comment]
Following up on related discussion in https://github.com/LibreTime/libretime/issues/935

Should everybody with any sort of availability focus on getting 3.0 out the door [before the API](https://discourse.libretime.org/t/planning-out-an-api/341/8)?

**Comment by paddatrapper** [comment]
That would be my recommendation. I'm stalled on the work I can do for 3.0 as my skills are python, and waiting on #951. That needs testing and merging before I can do more work there

**Comment by gusaus** [comment]
Are any other maintainers or contributors able to work with @paddatrapper on resolving 3.0 blockers? 
https://github.com/LibreTime/libretime/projects/2

... or is this 3.0 beta?

**Comment by xabispacebiker** [comment]
I can help on a few days..

**Comment by Robbt** [comment]
Yeah at this point we just need to do some house keeping on the various issues outlined. I still think that they are all a pretty good indicator of beta level vs alpha level although it would be easy enough for someone to review one and suggest that a fix isn't needed.

**Comment by gusaus** [comment]
> we just need to do some house keeping on the various issues outlined.

Based on activity here and [in the forums](https://discourse.libretime.org/), it seems like more folks are using and interested in LibreTime. There may even be some additional folks with time to contribute.

If one person (ideally not one of the core maintainers) could assume the PM/scrum master role, possibly we could coordinate some virtual sprints with the goal of 3.0.

Thoughts?

**Comment by gusaus** [comment]
Cross-posting this forum discussion posted by @Robbt https://discourse.libretime.org/t/libretime-3-beta-todos/529

If the release blockers are up to date, it looks like only a few tasks are not yet assigned - https://github.com/LibreTime/libretime/issues?q=is%3Aissue+is%3Aopen+label%3A3.0-release-blocker

Are there others that need to be claimed and resolved before we get beta 3 out the door? Are there particular tasks where folks claiming issues still need help?

**Comment by zklosko** [comment]
Since #953 was merged, @Robbt can you please check it off for clarity?

**Comment by Robbt** [comment]
What do you want me to check off ?

**Comment by zklosko** [comment]
You got it 👍

**Comment by paddatrapper** [comment]
For the easy install point, the Debian packages in the PPA are working for both Ubuntu Bionic and Xenial. We seem to have lost the docs for it at some point though

**Comment by zklosko** [comment]
@paddatrapper When trying on Ubuntu Bionic, I got this error

```
postinst called with unknown argument `install'
dpkg: error processing archive /home/ubuntu/libretime_3.0.0.alpha.8-1ubuntu3_amd64.deb (--unpack):
 new libretime package pre-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 /tmp/apt-dpkg-install-6iDZnQ/362-libretime_3.0.0.alpha.8-1ubuntu3_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by paddatrapper** [comment]
That was fixed in 3.0.0~alpha.8-1ubuntu4

**Comment by zklosko** [comment]
@paddatrapper I don't see that release uploaded to our Github repo.

**Comment by hairmare** [comment]
I amended the initial issue with a link to the [beta 3 project board](https://github.com/LibreTime/libretime/projects/2).

**Comment by paddatrapper** [comment]
Based on the discussions we've had in the watch file issue and the community meetings, I don't think the watch folder feature  is required for 3.0. We have an import script that people can run using cron if they want. It is an incredibly complicated feature to do porperly. So I think the only thing is the Docker support that is still outstanding.

@jooola is there anything you would like to include as a blocker for 3.0? Not sure how much of the UI re-write, architecture restructure, etc we want to block on? Also we need to define our support format - release support life cycles, format, etc

**Comment by jooola** [comment]
From my perspective:

- I would like #1378 and any related topic to be resolved, as we should do this once and for all (not having to rename the config file for example).
- The installer #1543 should be rewritten and the docs updated accordingly. This would mean:
  - the django migrations rewrite (we need a proper way to create migrations that are future proof, and I prefer to break everything now than rely on the php based migration tool)
  - Update the docs to follow the new installer: Install Libretime, dump a config sample, ask users to edit the config file, run the migrations.
  - The PRs that moves some settings stored in DB to the configuration file should be allow us to remove the web installer, and simplify the installation process as described above.
- The documentation should be updated to remove deprecated features.

I don't think we should consider shipping any of the UI and architectural changes, while it can be tempting to simplify some of the task we try to resolve before releasing v3.

**Comment by paddatrapper** [comment]
Ah, didn't realise it hasn't been merged yet, but the bulk import in #1659

**Comment by paddatrapper** [comment]
Yeah, I think that is reasonable. I haven't looked at the changes in #1543 yet, but will do

**Comment by jooola** [comment]
I am unsure about the bulk-import tool, it isn't really clean. If we ship it is should be considered as "temporary". I would prefer a scan storage task in the worker that could make a diff between the filesystem and the database, using the file path and md5 and then trigger the analyzer for any new file.

For this I would prefer having the worker inside django so we can mutate the database with the django ORM. But this involves moving the worker inside the django app.

**Comment by jooola** [comment]
Some days ago I send you my "roadmap" of pull request, as many pull request actually depends on other PRs.
- #1626
- #1644
- #1645
- #1543
- #1492

**Comment by paddatrapper** [comment]
For sure, an actual watch folder task would be great, but at this point I think incremental improvement is the way to go. We've been trying to achieve the perfect solution for everything while leaving our users waiting in limbo. Software is incremental, so we should be comfortable with implementing something and then improving or changing it later on

**Comment by jooola** [comment]
> For sure, an actual watch folder task would be great, but at this point I think incremental improvement is the way to go. We've been trying to achieve the perfect solution for everything while leaving our users waiting in limbo. Software is incremental, so we should be comfortable with implementing something and then improving or changing it later on

I don't necessarily want to delay, nor rush the release of v3, but once all my PR above are merged, I could focus on working on that, I just don't want to have those annoying things I am currently working on in the way while I try to implement this storage scanner. A dir scanner doesn't seem a huge amount of work.

But maybe you are right, you could release v3 and add this storage scanner along the way, as I don't see it as another breaking change. (according to our previous discussion on versioning and releasing, I agree we should not really care too much about breaking change and simply bump major versions, but this could also become a maintenance burden to sysadmin having to upgrade a major release every 3 month.)

**Comment by paddatrapper** [comment]
Personally, I would prefer prioritising the new UI over watch folders, but either way I agree it isn't a breaking change. Having every release be breaking would be bad, but that's where planning comes in - lump breaking changes together every few releases (which at our current release rate would be every couple years). We introduce a breaking change every two years anyway by dropping EOL Debian/Ubuntu versions.

I'm not trying to rush the release, but I am very concious that this issue alone has been open for 4 _years_... I would like to get things to a point where we have a set support structure and versioning. Predictability is more important to me than having a perfect product

**Comment by jooola** [comment]
> I'm not trying to rush the release, but I am very concious that this issue alone has been open for 4 years... I would like to get things to a point where we have a set support structure and versioning. Predictability is more important to me than having a perfect product

I think the issue couldn't be solved in 4 years because the project required a lot of cleanup, extra steps to be able to implement it cleanly. I feel now that we have:

- the django api with a way to create management commands
- a single storage path in the configuration file

we might have a chance to fix this easily.

**Comment by jooola** [comment]
I think we have a way to do bulk import with #1877, it will not be bullet proof though.

I'm tired of working with the legacy code, it isn't pleasant. Many of the issues below requires a good amount of work touching the php code.

I don't think the docker images will be ready for v3.0.0, it requires too effort just to split every service in a container, as we need to be able to configure each container locations and the legacy app heavily rely on the the same url to define this. See #1886 

I won't put too much effort into changing the password hashing algorithm either. See #1787 

The files track type field change to a constrained foreign key is also a lot of work touching to the legacy code, while I think it's dangerous to leave it like this, I don't want to break everything. Unless someone want to tackle this, feel free to jump in #1868 

The strangler fig pattern is interesting, but the amount of glue required to make it work with the legacy app is too much in my opinion. Maybe it could work better if we are able to break things, which should happen after v3. So I think we should release v3.

Anyway, I wonder what is missing in your opinion @paddatrapper ?

We could also either drop the milestones or the projects boards, as it seems like we have a duplicate way of tracking what is left to do for v3.

**Comment by paddatrapper** [comment]
> My struggle is to do the strangler fig pattern, while we push towards stability. As it would require a minimum of risky changes that we don't want to do right before cutting a beta release.

I think that replacing the UI will have to be after the beta release, yes. I think the priority should be getting this out

**Comment by jooola** [comment]
From the initial release blocker list the only one left is:
- https://github.com/libretime/libretime/issues/42 Line-in recording
 
But I think we can remove it from the list, and implement this after v3 is released. Using liquidsoap directly, but this requires to touch playout, which should be done after v3.

With that in mind all the v3 blocker seem completed, the milestone still references some really old issues than could probbaly be closed. I think we can release alpha.13 and next release will be the beta.0 !

**Comment by paddatrapper** [comment]
I agree with that sentiment. Though do we need an alpha.13 or should we just release the next version as beta.0?

**Comment by jooola** [comment]
I think we should release alpha.13 soon as it contain important fixes, and give ourselves a week or two to polish everything before releasing beta.0.

I would like to open some cleanup possibly breaking PRs, so those pre beta.0 weeks should give me enough time to do so.

This makes 2 release in 2 week, not sure if that too much. The beta.0 upgrade process will not be as big as the alpha.13 release, so maybe that not necessarily a problem.

**Comment by paddatrapper** [comment]
Sounds good. Maybe have 3 weeks between, as I'm at DebConf for the next two weeks, so will be a little erratically available. So do alpha.13 this week and beta.0 the week of Monday 25 July

**Comment by jooola** [comment]
We might also want to include the following before releasing beta.0:
- #1637
- #1719
- #1186

I have it running in production for few months now, and I didn't have any problem.

**Comment by jooola** [comment]
LibreTime v3.0.0-beta.0 is due in the coming weeks (after the 27/07/2022). The few remaining blockers have been rescheduled for future versions in the 3.x development series.

Closing as resolved.

**Comment by jooola** [comment]
The announced due date above is not valid anymore. We reconsidered the container feature, we want to ship docker support in v3 and it requires breaking changes. Thoses changes arent trivial, so it will probably requires some more weeks.
